src/cpp/ext/csm/metadata_exchange.cc

```diff
@@ -226,17 +229,34 @@ class MeshLabelsIterable : public LabelsIterable {
                                 struct_pb.struct_pb, kMetadataExchangeTypeKey,
                                 struct_pb.arena.ptr()));
     }
-    // Only handle GKE type for now.
+    if (pos_ == local_labels_size + 2) {
+      return std::make_pair(
+          kPeerCanonicalServiceAttribute,
+          GetStringValueFromUpbStruct(struct_pb.struct_pb,
+                                      kMetadataExchangeCanonicalServiceKey,
+                                      struct_pb.arena.ptr()));
+    }
     switch (remote_type_) {
```

Comment:
I think this could be written more clearly using some helper functions, so that we can use the same code for the fixed labels and for any remote type:

```
absl::Span<const RemoteAttribute> GetAttributesForType(GcpResourceType remote_type) {
  switch (remote_type) {
    case GcpResourceType::kGke:
      return kGkeAttributeList;
    case GcpResourceType::kGce:
      return kGceAttributeList;
    default:
      return {};
  }
}

static constexpr std::array<RemoteAttribute, 2> kFixedAttributes = {
    RemoteAttribute{kPeerTypeAttribute, kMetadataExchangeTypeKey},
    RemoteAttribute{kPeerCanonicalServiceAttribute,
                    kPeerCanonicalServiceAttribute},
};

absl::optional<std::pair<absl::string_view, absl::string_view>> NextFromAttributeList(
    const StructPb& struct_pb, absl::Span<const RemoteAttribute> attributes, size_t start_index) {
  const size_t index = pos_ - start_index;
  if (index >= attributes.size()) return absl::nullopt;
  ++pos_;
  const auto& attribute = attributes[index];
  return std::make_pair(
      attribute.otel_attribute,
      GetStringValueFromUpbStruct(
          struct_pb.struct_pb,
          attribute.metadata_attribute,
          struct_pb.arena.ptr()));  
}
```

Then the code here can handle basically everything except local labels without duplication:

```
const size_t fixed_attribute_end = local_labels_size + kFixedAttributes.size();
if (pos_ < fixed_attribute_end) {
  return NextFromAttributeList(struct_pb, kFixedAttributes, local_labels_size);
}
auto attributes = GetAttributesForType(remote_type_);
return NextFromAttributeList(struct_pb, attributes, fixed_attribute_end);
```

Reply:
Suggestion accepted. Thanks!